File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-10.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 1 or len(s) == 0:
            return len(s)

        left = 0
        right = 1
        seen = []
        seen.append(s[left])
        max_substring = 0
        while right < len(s):

            if s[left] != s[right] and s[right] not in seen:
                seen.append(s[right])
                right += 1
            else:
                max_substring = max(len(seen), max_substring)
                i = seen.index(s[right])
                seen = seen[i+1:]
                seen.append(s[right])
                left = right
                right += 1
        return max(max_substring, len(seen))

File: Data Structures & Algorithms/max-water-container/submission-10.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int left = 0, right = heights.size() - 1; 
        size_t max_area = 0;

        while (left < right){
            size_t max_height = std::min(heights[left], heights[right]);
            auto length = right - left;

            max_area = std::max(max_area, max_height * length);

            if (heights[left] < heights[right]){
                left++;
            } else {
                right--;
            }

        }

        return max_area;
    }
};
